### DISCUSSION TITLE
Priority Algorithms to be Implemented

### DISCUSSION THREAD

**Discussion by jejjohnson** [facebookincubator/flowtorch#87] [Q&A] [answered]
Hi, 

Firstly, I wanted to say thank you for releasing this! I’m hoping this will be a big help for the normalizing flow community.

I’m really hoping to contribute this year (2022). So I wanted to know what are some of the high priority algorithms that you all would like to have implemented. Some things that come to mind are some of the usual characters like the different variants of coupling transforms (masked, spline, mixture cdf), some other linear transforms (householder, 1x1 conv, sylvester, exponential), and some dimensionality changes (reshaping, augmentation, slicing). And then I’ve already seen some discussions about conditional flows. 

I’m not the best programmer so I’d rather wait to see some of the design choices you all make for some of the base classes before I start 😅

Thanks in advanced!

**Comment by stefanwebb** [comment, marked as answer]
Hi @jejjohnson,

Glad to hear that! The success of the library depends on potential contributors like yourself, so I'm excited to learn of your interest :)

The feedback I've gotten so far is that the most requested features are:
* conditional normalizing flows
* coupling bijectors
* continuous normalizing flows

The first two require some discussion about the API within the core developer team, although continuous normalizing flows would be a fantastic task for a contributor to take on!

Alternatively, it would be fantastic to get some help migrating the code for Householder, 1x1 conv, Sylvester bijectors etc. from Pyro if that would be more interesting to you :)

Another item that would be ideal for contributors would be to write tutorials on how to do useful things with normalizing flows, replicating results from applied papers. Soon, we will have a system that displays Python tutorial notebooks on the website as well as linking to a Colab.

Let me know how I can help!

**Reply by jejjohnson** [reply]
Sounds good! I think I will start looking at some of the pyro modules that can be imported into flowtorch. It would be a good start at my level to get myself familiar with the API. And I will skip the coupling transforms and conditional stuff for now (although I'm super eager to see the API you all finally come up with in the end).

In addition, I'm a big fan of the [nflows](https://github.com/bayesiains/nflows) package and an even bigger fan of the authors of code [survae-flows](https://github.com/didriknielsen/survae_flows). I'll look to those for inspiration as well.

I will be in touch!
